include/deal.II/base/template_constraints.h

```diff
@@ -25,10 +25,26 @@
 
 DEAL_II_NAMESPACE_OPEN
 
-namespace
+namespace internal
```

Comment:
why don't you like the anonymous namespace here?

Reply:
i think @drwells was recently fixing those, i.e. adding `internal` to anonymous namespaces. I forgot the reason, though :smile:

Reply:
Anonymous namespaces in header files are generally a not so great idea.

An anonymous namespace is equivalent to a named namespace where the name is automatically created for each "translation unit" (i.e., .cc file that `#include`s the header). That means that a symbol in an anonymous namespace in a header file is *distinct* between two object files. Most of the time this has no practical impact other than the fact that the linker cannot, for example, unify the code for one function. But you'd get into trouble if you had, for example, a global variable in such a namespace, or maybe a `static` variable in a function in such a namespace: these variables are not in fact global, but duplicated in every .cc file that `#include`s the header file.

For such reasons, it's generally not a great idea to use anonymous namespaces in header files. Most of the time, nothing bad happens, but when something bad happens, it's a really bizarre situation that difficult to debug.